app/src/main/java/com/hedvig/app/authenticate/insurely/InsurelyAuthViewModel.kt

```diff
@@ -0,0 +1,53 @@
+package com.hedvig.app.authenticate.insurely
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class InsurelyAuthViewModel(
+    private val reference: String,
+    private val getDataCollectionUseCase: GetDataCollectionUseCase
+) : ViewModel() {
+
+    private val _viewState = MutableStateFlow(ViewState())
+    val viewState: StateFlow<ViewState>
+        get() = _viewState
+
+    data class ViewState(
+        val autoStartToken: String? = null,
+        val authStatus: DataCollectionResult.Success.CollectionStatus? = null,
+        val error: DataCollectionResult.Error? = null,
+    )
+
+    init {
+        viewModelScope.launch {
+            getDataCollectionUseCase.getCollectionStatus(reference).collect { result ->
```

Comment:
I was thinking something about these cases.
Since the we launch this on the viewModelScope, on a hot StateFlow, this would still be actively listening to the subscription even if there are no subscribers to it right?

Inspired from [something I read a while back](https://medium.com/androiddevelopers/migrating-from-livedata-to-kotlins-flow-379292f419fb) under the section "The WhileSubscribed strategy" the entire class could also look like this:
```kotlin
class InsurelyAuthViewModel(
    reference: String,
    getDataCollectionUseCase: SubscribeToDataCollectionUseCase,
) : ViewModel() {

    val viewState: StateFlow<ViewState> = getDataCollectionUseCase
        .getCollectionStatus(reference)
        .map { result ->
            when (result) {
                is DataCollectionResult.Error -> ViewState.Error
                is DataCollectionResult.Success.NorwegianBankId -> ViewState.Success(
                    autoStartToken = result.norwegianBankIdWords,
                    authStatus = result.status,
                )
                is DataCollectionResult.Success.SwedishBankId -> ViewState.Success(
                    autoStartToken = result.autoStartToken,
                    authStatus = result.status,
                )
            }
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = ViewState.Initial,
        )


    sealed class ViewState {
        object Initial : ViewState() // Since I saw the the initial was just the model with all parameters being null maybe a plain "Initial" state is adequate?
        object Error : ViewState()
        data class Success(
            val autoStartToken: String?,
            val authStatus: DataCollectionResult.Success.CollectionStatus,
        ) : ViewState()
    }
}
```
With the extra benefit that the class is now even smaller.

Any opinions on this? @hedvigoscar you too.

Reply:
Good points, and it looks good, I think.

Reply:
Agree, looks 👌